lib/ngx/re.md

```diff
@@ -0,0 +1,183 @@
+
+Table of Contents
+=================
+
+* [Name](#name)
+* [Status](#status)
+* [Synopsis](#synopsis)
+* [Description](#description)
+* [Methods](#methods)
+    * [split](#split)
+* [Community](#community)
+    * [English Mailing List](#english-mailing-list)
+    * [Chinese Mailing List](#chinese-mailing-list)
+* [Bugs and Patches](#bugs-and-patches)
+* [Author](#author)
+* [Copyright and License](#copyright-and-license)
+* [See Also](#see-also)
+
+Status
+======
+
+This Lua module is currently considered experimental.
+
+Synopsis
+========
+
+```lua
+local ngx_re = require "ngx.re"
+
+-- split
+local res, err = ngx_re.split("a,b,c,d", ",")
+--> res is now {"a", "b", "c", "d"}
+```
+
+[Back to TOC](#table-of-contents)
+
+Description
+===========
+
+This Lua module provides a Lua API which implements convenience utilities for
+the `ngx.re` API.
+
+[Back to TOC](#table-of-contents)
+
+Methods
+=======
+
+All the methods of this module are static (or module-level). That is, you do
+not need an object (or instance) to call these methods.
+
+[Back to TOC](#table-of-contents)
+
+split
+----------------
+**syntax:** *res, err = ngx_re.split(subject, regex, options?, ctx?, max?, res?)*
+
+Splits the `subject` string using the Perl compatible regular expression
+`regex` with the optional `options`.
+
+This function returns a Lua (array) table (with integer keys) containing the
+splitted values.
+
+In case of error, `nil` will be returned as well as a string describing the
+error.
+
+When `regex` contains a sub-match capturing group, and when such a match is
+found, the first submatch capture will be inserted in between each splitted
+value, like so:
+
+```lua
+local ngx_re = require "ngx.re"
+
+local res, err = ngx_re.split("a,b,c,d", "(,)")
+-- res is now {"a", ",", "b", ",", "c", ",", "d"}
+```
+
+The optional `ctx` table argument can be a Lua table holding an optional `pos`
+field. When the `pos` field in the `ctx` table argument is specified,
+`ngx_re.split` will start splitting the `subject` from that index:
+
+```lua
+local ngx_re = require "ngx.re"
+
+local res, err = ngx_re.split("a,b,c,d", ",", nil, {pos = 5})
+-- res is now {"c", "d"}
+```
+
+The optional `max` argument is a number that when specified, will prevent
+`ngx_re.split` from adding more than `max` matches to the `res` array:
+
+```lua
+local ngx_re = require "ngx.re"
+
+local res, err = ngx_re.split("a,b,c,d", ",", nil, nil, 3)
+-- res is now {"a", "b", "c,d"}
+```
+
+Specifying `max <= 0` disables this behavior, meaning that the number of
+results won't be limited.
+
+The optional 6th argument `res` can be a table that `ngx_re.split` will re-use
+to hold the results instead of creating a new one, which can improve
+performance in hot code paths. It is used like so:
+
+```lua
+local ngx_re = require "ngx.re"
+
+local my_table = {"hello world"}
+
+local res, err = ngx_re.split("a,b,c,d", ",", nil, nil, nil, my_table)
+-- res/my_table is now {"a", "b", "c", "d"}
+```
+
```

Comment:
Better add a note regarding the user responsibility of clearing this `res` table and our behavior of setting a trailing `nil` value as the guard?